### PR TITLE
made changes for heap

### PR DESCRIPTION
**Pull request by ruchiramoitra**

### CODE REVIEW

File: v0/destinations/heap/transform.js [hunk 1]
```diff
@@ -39,20 +83,32 @@ function commonPayload(message, rawPayload, type) {
...
+  let email;
   switch (type) {
     case EventType.TRACK:
       propsArray = get(message, "properties")
         ? Object.keys(message.properties)
         : null;
       rudderPropertiesObj = message.properties;
-      rawPayload.identity = message.context.traits.email;
+      email =
+        (message.context.traits ? email : null) || message.properties.email;
```
Comment (sayan-mitra) [inline v0/destinations/heap/transform.js:94]: (message.context.traits ? email : null)  ==> email is undefined here right?
so rawPayload.identity = null is valid but rawPayload.identity = undefined (previous version) is not?

File: v0/destinations/heap/transform.js [hunk 2]
```diff
@@ -39,20 +83,32 @@ function commonPayload(message, rawPayload, type) {
...
+      else
+        throw Error("Email for identity is required for sending track calls.");
       break;
     case EventType.IDENTIFY:
       propsArray = get(message.context, "traits")
         ? Object.keys(message.context.traits)
         : null;
+      email =
+        message.context.traits.email ||
+        (message.properties ? message.properties.email : null);
```
Comment (sayan-mitra) [inline v0/destinations/heap/transform.js:105]: will identify have message.properties?

File: v0/destinations/heap/transform.js [hunk 3]
```diff
@@ -7,7 +7,51 @@ const {
   defaultRequestConfig
 } = require("../util");
 
-function responseBuilder(payload, message, heapConfig) {
+function flatten(target, opts) {
```
Comment (sayan-mitra) [inline v0/destinations/heap/transform.js:10]: heap needs  flattened properties?
we are not using opts, can this be removed? Also, move this method to utils?
can we have a description and an example of what it does with javascript object, object array and array types

Comment (ruchiramoitra) [reply v0/destinations/heap/transform.js:10]: this function is also used in another destination Kissmetrics. there opts is used so making it generic and keeping it in utils also adding comments.

File: v0/destinations/heap/transform.js (hunk 2 above)
Comment (ruchiramoitra) [reply v0/destinations/heap/transform.js:105]: properties can be sent through identify calls right? 
Thus checking once and then assigning.

File: v0/destinations/heap/transform.js (hunk 3 above)
Comment (ruchiramoitra) [reply v0/destinations/heap/transform.js:10]: Yes heap was not accepting nested properties.

File: v0/destinations/heap/transform.js (hunk 2 above)
Comment (sayan-mitra) [reply v0/destinations/heap/transform.js:105]: do we have that provision in our sdks?

File: v0/destinations/heap/transform.js (hunk 1 above)
Comment (ruchiramoitra) [reply v0/destinations/heap/transform.js:94]: changed the logic

**Comment by sayan-mitra** [comment]
👍

**Pull request merged**